Add tests for the suggest method

teytaud opened the pull request:
## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Motivation and Context / Related issue

We want to be sure that "suggest" works.
It turns out that it does work.

## How Has This Been Tested (if it applies)

<!--- Please describe here how your modifications have been tested. -->

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] The documentation is up-to-date with the changes I made.
- [x] I have read the [**CONTRIBUTING**](https://facebookresearch.github.io/nevergrad/contributing.html) document and completed the CLA (see [**CLA**](https://facebookresearch.github.io/nevergrad/contributing.html#contributor-license-agreement-cla)).
- [x] All tests passed, and additional code has been covered with new tests.

<!--- In any case, don't hesitate to join and ask questions if you need on Nevergrad users Facebook group https://www.facebook.com/groups/nevergradusers/ -->

jrapin commented on nevergrad/optimization/test_optimizerlib.py:
```diff
@@ -197,6 +197,86 @@ def test_infnan(name: str) -> None:
 
+def suggestable(name: str) -> bool:
+    # Some methods are not good with suggestions.
+    if "TBPSA" in name:
+        return False
+    if "EMNA" in name or "EDA" in name:
+        return False
+    if "Stupid" in name:
+        return False
+    if "BO" in name:
+        return False
+    if "Pymoo" in name:
+        return False
+    return True
```
```suggestion
    keywords = ["TBPSA", "BO", "EMNA", "EDA", "BO", "Stupid", "Pymoo"]
    return not any(x in name for x in keywords)
```

I still don't like relying on names though, would be better to use the base classes, or rely on TellNotAskedNotSupported

jrapin commented on nevergrad/optimization/test_optimizerlib.py:
```diff
@@ -197,6 +197,86 @@ def test_infnan(name: str) -> None:
+    target = lambda x: 0 if np.all(np.asarray(x, dtype=int) == xs) else 1
+    optim.suggest(xs)
+    optim.minimize(target)
+    assert np.all(optim.recommend().value == xs), "{name} proposes {optim.recommend().value} instead of {xs}"
```
```suggestion
    assert not target(optim.recommend().value), "{name} proposes {optim.recommend().value} instead of {xs}"
```

teytaud replied:
I don't get it, your version does not have xs anywhere in the equation ?

teytaud replied:
Oh I see, you want to use the fact that the optimum has value 0.

jrapin reviewed (changes requested):
There is a lot of **very** redundant code, you could easily change the target depending on the case instead

teytaud commented:
> There is a lot of **very** redundant code, you could easily change the target depending on the case instead

The different tests are distinct though. Some cases are continuous and others are discrete. Also, in some cases we just check "if the optimum has been suggested then we find it" and in other cases we check the stronger requirement "if a good suboptimal point has been provided then we find the optimum".

teytaud commented:
> > There is a lot of **very** redundant code, you could easily change the target depending on the case instead
> 
> The different tests are distinct though. Some cases are continuous and others are discrete. Also, in some cases we just check "if the optimum has been suggested then we find it" and in other cases we check the stronger requirement "if a good suboptimal point has been provided then we find the optimum".

I think compacting more would decrease readability.

jrapin commented:
> > > There is a lot of **very** redundant code, you could easily change the target depending on the case instead
> > 
> > 
> > The different tests are distinct though. Some cases are continuous and others are discrete. Also, in some cases we just check "if the optimum has been suggested then we find it" and in other cases we check the stronger requirement "if a good suboptimal point has been provided then we find the optimum".
> 
> I think compacting more would decrease readability.

But increase cost of maintenance, there is too much shared code here, I am sure it's possible to be both readable and factorized

teytaud commented:
> > > > There is a lot of **very** redundant code, you could easily change the target depending on the case instead
> > > 
> > > 
> > > The different tests are distinct though. Some cases are continuous and others are discrete. Also, in some cases we just check "if the optimum has been suggested then we find it" and in other cases we check the stronger requirement "if a good suboptimal point has been provided then we find the optimum".
> > 
> > 
> > I think compacting more would decrease readability.
> 
> But increase cost of maintenance, there is too much shared code here, I am sure it's possible to be both readable and factorized

ok for you now ?

jrapin commented on nevergrad/optimization/test_optimizerlib.py:
```diff
@@ -196,6 +196,90 @@ def test_infnan(name: str) -> None:
+
+def suggestion_testing(
+    name: str,
+    instrumentation: tp.Union[ng.p.Array, ng.p.Instrumentation],
```
```suggestion
    instrumentation: np.p.Array,
```
you only use Array, and the tests will fail with anything else anyway

jrapin commented on nevergrad/optimization/test_optimizerlib.py:
```diff
@@ -196,6 +196,90 @@ def test_infnan(name: str) -> None:
+    suggestion_testing(name, instrum, suggestion, 1500, target, optimum)
+
+
+def good_at_c0_suggest(r: str) -> bool:
+    return "NGOpt" == r  # Let us check on NGOpt only for now.
```
is that really needed as a function?

teytaud replied:
Removed.

jrapin reviewed (commented):
given it's added 3 very specific tests on one particular topic, it may be better to move to a new file test_suggest.py

jrapin reviewed (commented):
the code seems very redundant, but not sure it could be restructured better

teytaud commented:
> the code seems very redundant, but not sure it could be restructured better

is it ok for you now ? i refactored cycling.py but not the underlying computations: i assume that original authors would be lost if I changed the deep stuff.

jrapin reviewed (approved):
I would still move all suggestions tests to one test_suggest.py file because we start to have quite a few of them, and test_optimizerlib is too big
